fix invalidations in sort! from Static.jl

ranocha opened the pull request:
This should hopefully fix quite some invalidations coming from Static.jl.

<details>

<summary>
Here is the code:

</summary>

```julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0 (2022-08-17)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(@v1.8) pkg> activate --temp

(jl_PDrBpd) pkg> add Static
    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
    Updating `/tmp/jl_PDrBpd/Project.toml`
  [aedffcd0] + Static v0.7.6
    Updating `/tmp/jl_PDrBpd/Manifest.toml`
  [615f187c] + IfElse v0.1.1
  [aedffcd0] + Static v0.7.6

julia> using SnoopCompileCore

julia> invalidations = @snoopr using Static

julia> using SnoopCompile

julia> trees = invalidation_trees(invalidations)
7-element Vector{SnoopCompile.MethodInvalidations}:
...
 inserting !(::False) in Static at ~/.julia/packages/Static/sVI3g/src/Static.jl:427 invalidated:
...
                 160: signature Tuple{typeof(!), Any} triggered MethodInstance for sort!(::Vector{String}, ::Int64, ::Int64, ::Base.Sort.InsertionSortAlg, ::Base.Order.Ordering) (127 children)
```

</details>

Since `!lt(...)` is used here in an `if` stament - and `if` statements only accept `Bool`s -
it appears to be sane to assert that `lt` returns a `Bool` here.

See also https://github.com/JuliaLang/julia/pull/46481, https://github.com/JuliaLang/julia/pull/46490

ranocha committed:
fix invalidations in sort! from Static.jl

Tokazama commented:
Another instance where it might be easier in the long run to annotate the function definition itself?

```julia
lt(::Left, x::T, y::T)::Bool where {T<:Floats} = slt_int(y, x)
lt(::Right, x::T, y::T)::Bool where {T<:Floats} = slt_int(x, y)
```

ranocha committed:
move Bool annotation to definition of lt

ranocha commented:
I'm not sure whether it works and fixes the invalidations but since you seem to be sure I did what you suggested in [bbf1bde](https://github.com/JuliaLang/julia/pull/46491/commits/bbf1bdeef6b451a4d2e092678953301f8a277961)

Tokazama commented:
You can ensure that these fix invalidations in master by building from you branch of Julia and looking at the invalidations again. Depending on compiler work this may not be reflective of what happens for 1.8

ranocha committed:
return type assertion with where needs long form

ranocha commented:
There do not seem to be invalidations anymore on recent `master` with this PR on top (after moving the type assertions to the definitions of `lt` as suggested by @Tokazama).

ranocha committed:
Revert "return type assertion with where needs long form"

This reverts commit 6b218f8a8b813ed77551ae64f13b15f4b4a8db75.
The new form did not fix invalidations in Julia 1.8.

ranocha committed:
Revert "move Bool annotation to definition of lt"

This reverts commit 0de5fe58b1e4450ea44619c6542f93eeb12c3fb0.

ranocha commented:
The version proposed by @Tokazama does not fix invalidations from a clean build of Julia from the `release-1.8` branch with this PR on top of it (on my system).  However, the version I implemented at first did fix the invalidations. Thus, I reverted the change proposed by @Tokazama.

timholy commented:
@Tokazama, inference stops checking whether all possible callees return `Bool` if the number of methods exceeds 3. Thus in many cases, just adding method annotations does nothing to fix invalidations.

Tokazama commented:
> @Tokazama, inference stops checking whether all possible callees return `Bool` if the number of methods exceeds 3. Thus in many cases, just adding method annotations does nothing to fix invalidations.

That's really interesting because one of the ideas being thrown around to resolve invalidations more proactively by Stefan was to have some way of declaring all method variants inherit a type assertion (e.g., `function isequal::Bool end` would then require all methods return a `Bool` or error even if they don't have the type assert).

timholy commented:
#42372, #36463

timholy reviewed (approved):
Since this calls `isless`, this seems perfectly safe.
